Cookbook/Proposals/FeatureSwitchesRedesign.md

```diff
@@ -0,0 +1,218 @@
+
+Inside the module we can extend `World` with a module specific namespace and keep its feature switches and configurations separate from the rest of `World` properties.
+
+```swift
+// In the feature framework
+extension Current {
+    private static private(set) var someFeature: SomeFeatureModule!
+    
+    // way for the app to set static configurations for the feature
+    public func configureSomeFeature(
+        config: SomeFeatureModule.Configuration
+    ) {
+        World.someFeature = SomeFeatureModel(config: config)
+    }
+    
+    public var someFeature: FeatureModule {
+        return World.someFeature
+    }
+    
+    public struct SomeFeatureModule {
+       
+        public let featureSwitches: FeatureSwitches
+    
+        // dynamic feature switches
+        public struct FeatureSwitches {
+            public let isNewFeatureEnabled = RemoteFeatureSwitch<Bool>(
+                name: “New feature”,
+                key: “is_new_feature_enabled”
+            )
+            public let inProgressFeatureEnabled = LocalFeatureSwitch<Bool>(
+                name: “Another feature”,
+                key: “is_another_feature_enabled”
+            )
+        }
+        
+        // static configurations
+        public let config: Configuration
+        
+        public struct Configuration {
+            public let someAppSpecificConfiguration: Bool // can have a default value
+        }
+    }
+}
+```
+
+This way feature switches will be accessed in the feature builders via `Current.someFeature.featureSwitches` or `Current.someFeature.config`.
+
+If the type of the switch changes (i.e. from local to remote switch) we change it in a single place where it is defined rather in every place where it is used. To abstract away difference between static configuration and dynamic switches we can use a key-path based getters or subscripts:
+
+```swift
+extension World.SomeFeatureModule {
+    public func get<T>(_ keyPath: KeyPath<RemoteFeatureSwitch<T>, World.SomeFeatureModule.FeatureSwitches>) -> T {
+        return featureSwitches.keyPath[keyPath].value
+    }
+    
+    public func get<T>(_ keyPath: KeyPath<T, World.SomeFeatureModule.Configuration>) -> T {
+        return config.keyPath[keyPath]
+    }
+}
+```
+
+With Swift 5.1 we can improve this a bit with property wrappers which will allow us to use `isNewFeatureEnabled` directly rather than with `isNewFeatureEnabled.value`:
+
+```swift
+@RemoteFeatureSwitch(
+    name: “New feature”,
+    key: “is_new_feature_enabled”
+)
+var isNewFeatureEnabled: Bool
+```
+
+Also we will be able to use `dynamicMemberLookup` with keypaths instead of custom getters or subscripts so that we wil be able to access dynamic feature switches or static configuration as `Current.someFeature.isNewFeatureEnabled` instead of `Current.someFeatures.get(\.isNewFeatureEnabled)`
```

Comment:
```suggestion
Also we will be able to use `dynamicMemberLookup` with keypaths instead of custom getters or subscripts so that we will be able to access dynamic feature switches or static configuration as `Current.someFeature.isNewFeatureEnabled` instead of `Current.someFeatures.get(\.isNewFeatureEnabled)`
```